pkg/systemd/overview-cards/usageCard.jsx

```diff
@@ -29,59 +29,91 @@ import "./usageCard.less";
 
 const _ = cockpit.gettext;
 
-const UPDATE_DELAY = 5000;
+const METRICS_SPEC = {
+    payload: "metrics1",
+    source: "internal",
+    interval: 5000,
+    metrics: [
+        { name: "cpu.basic.user", derive: "rate" },
+        { name: "cpu.basic.system", derive: "rate" },
+        { name: "cpu.basic.nice", derive: "rate" },
+        { name: "memory.used" },
+    ]
+};
 
 export class UsageCard extends React.Component {
     constructor(props) {
         super(props);
 
-        this.state = { pollingEnabled: true };
-        this.updateMemoryInfo = this.updateMemoryInfo.bind(this);
-    }
+        this.metrics_channel = null;
 
-    componentDidMount() {
-        this.updateMemoryInfo();
+        this.state = { memTotal: 0, memUsed: 0, numCpu: 1, cpuUsed: 0 };
 
-        cockpit.addEventListener("visibilitychange", () => {
-            this.setState((prevState, _) => ({ pollingEnabled: !prevState.pollingEnabled }));
-        }, () => this.updateMemoryInfo());
-    }
+        machine_info.cpu_ram_info().done(info => this.setState({
+            memTotal: Number((info.memory / (1024 * 1024 * 1024)).toFixed(1)),
+            numCpu: info.cpus,
+        }));
+
+        this.onVisibilityChange = this.onVisibilityChange.bind(this);
+        this.onMetricsUpdate = this.onMetricsUpdate.bind(this);
 
-    componentWillUnmount() {
-        this.setState({ pollingEnabled: false });
+        cockpit.addEventListener("visibilitychange", this.onVisibilityChange);
+        this.onVisibilityChange();
     }
 
-    updateMemoryInfo() {
-        if (!this.state.pollingEnabled)
+    onVisibilityChange() {
+        if (cockpit.hidden && this.metrics_channel !== null) {
+            this.metrics_channel.removeEventListener("message", this.onMetricsUpdate);
+            this.metrics_channel.close();
+            this.metrics_channel = null;
             return;
+        }
 
-        machine_info.cpu_ram_info().done(info => {
-            this.setState({
-                memTotal: Number((info.memory / (1024 * 1024 * 1024)).toFixed(1)),
-                memAvailable: Number((info.available_memory / (1024 * 1024 * 1024)).toFixed(1))
-            });
-        });
-        window.setTimeout(this.updateMemoryInfo.bind(this), UPDATE_DELAY);
+        if (!cockpit.hidden && this.metrics_channel === null) {
+            this.metrics_channel = cockpit.channel(METRICS_SPEC);
+            this.metrics_channel.addEventListener("closed", (ev, error) => console.error("metrics closed:", error));
+            this.metrics_channel.addEventListener("message", this.onMetricsUpdate);
+        }
+    }
+
+    onMetricsUpdate(event, message) {
+        const data = JSON.parse(message);
+        // ignore meta messages
+        if (!Array.isArray(data))
+            return;
+        // CPU metrics are in ms/s; divide by 10 to get percentage
+        const cpu = Math.round((data[0][0] + data[0][1] + data[0][2]) / 10 / this.state.numCpu);
```

Comment:
Because of `derive: rate`, the first sample for CPU usage will always be false.  In that case, we should keep the old value.  JS being JS, false + false + false == 0 so this just gives zero when the page loads until we have the second sample.  However, this also happens when visibility changes: Everytime one returns to the Overview, CPU usage is zero for 5 seconds.  It would be a bit less wrong to show the value that we had when the page became invisible.

(Or we hide the whole display until we have real data.)

Reply:
Thinking out loud: We could add a new feature to the metrics channel where we would have separate intervals for reporting and for computing the rate.  For example, we could use an interval of 200ms for computing the rate, but only once every 5 seconds.  Then the first useful sample would already come after 200ms.  (And the measurements would be close to the true instantaenous rate, instead of a 5 second average.)  Hmm.

Reply:
You did the `false` handling in the fixup, thanks!